Topics/Tech_Stacks/Pythonic-Iteration-Patterns.md

```diff
@@ -0,0 +1,238 @@
+
+
+### Dictionaries
+
+Iterating over a dictionary means iterating over the _keys_:
+```python
+for key in {'a': 3, 'b': 1, 'c': 4}:
+    print(key)  # prints 'a', then 'b', then 'c'
+
+# Alternatively:
+for key in {'a': 3, 'b': 1, 'c': 4}.keys():
+    print(key)
+```
+To iterate over the _values_, use the `dict.values` method:
+```python
+for value in {'a': 3, 'b': 1, 'c': 4}.values():
+    print(value)  # prints '3', then '1', then '4'
+```
+To iterate over both keys _and_ values, use the `dict.items` method:
+```python
+for key, value in {'a': 3, 'b': 1, 'c': 4}.items():
+    print(f'{key} -> {value}')  # prints 'a -> 3', then 'b -> 1', then 'c -> 4'
+```
+[Starting in Python 3.7, iterating over a dict is guaranteed to produce items in insertion order.](
+https://stackoverflow.com/questions/39980323/are-dictionaries-ordered-in-python-3-6)
+
+
+### Sets
+
+Lastly, you can also iterate over a set, which produces items in an unspecified order:
+```python
+for element in {'a', 'b', 'c'}:
+    print(element)  # prints 'a', 'b', and 'c' in some order
+```
+
+
+
+## Other tools to look out `for`
+
+As a rule of thumb, if you find yourself writing code that looks something like this:
+```python
+for index in indices_of_a_collection:
+    do_something_with(index, collection[index])
+```
+...there's probably a more idiomatic way to go about things.
+
+Here are a few of those potential ways to go about things:
+
+
+### `enumerate`
+```python
+lst = [3, 1, 4]
+
+# Instead of:
+for i in range(len(lst)):
+    print(f'{i}: {lst[i]}')
+
+# Write:
+for i, x in enumerate(lst):
+    print(f'{i}: {x}')  # prints '0: 3', then '1: 1', '2: 4'
+```
+`enumerate` also has an optional `start` argument that allows for counting from numbers other than 0:
+```python
+for i, char in enumerate('abc', start=1):
+    print(f'{i}: {char}')  # prints '1: a', then '2: b', then '3: c'
+```
+
+
+### `zip`
+```python
+lst_a = [1, 5, 9]
+lst_b = [2, 6, 5]
+lst_c = [3, 5, 8]
+
+# Instead of:
+for i in range(len(lst_a)):
+    print(f'{lst_a[i]} & {lst_b[i]} & {lst_c[i]}')
+
+# Write:
+for a, b, c in zip(lst_a, lst_b, lst_c):
+    print(f'{a} & {b} & {c}')  # prints '1 & 2 & 3', then '5 & 6 & 5', then '9 & 5 & 8'
+
+# And instead of:
+for i in range(len(lst_a)):
+    print(f'{i}: {lst_a[i]} & {lst_b[i]} & {lst_c[i]}')
+
+# Write:
+for i, (a, b, c) in enumerate(zip(lst_a, lst_b, lst_c)):
+    print(f'{i}: {a} & {b} & {c}')  # prints '0: 1 & 2 & 3', then '1: 5 & 6 & 5', then '2: 9 & 5 & 8'
+```
+Note that this is only equivalent if all the arguments to `zip` have the same number of items to iterate over.
+By default, iteration cuts off when any of the arguments to `zip` run out of items to iterate over.
+`zip` also supports an optional `strict` keyword argument to raise an error
+if the given iterables do not have the same number of items to iterate over.
+Also see [`zip_longest` from the `itertools` module](
+https://docs.python.org/3/library/itertools.html#itertools.zip_longest).

```

Comment:
This example is duplicate because you've provided the examples for enumerate above and this is irrelevant to zip. If you want to show that the return object of zip (i.e., iterator of tuples) can also be applied to enumerate, it's better to explicilty show the complete input/return types of these built-in functions.
Also, for the 'default' behavior and 'strict' keyword, it's better to provide 2 more examples. (I find the examples in the python documentation easy to follow. https://docs.python.org/3/library/functions.html?highlight=enumerate#zip)